### PR TITLE
feat: testservice sends link previews (WPB-9877) (WPB-2652)

### PR DESCRIPTION
**Pull request by mythsunwind**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9877" title="WPB-9877" target="_blank">TaskWPB-9877</a>  [Kalium] Link preview persistence and handling of proto messages
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This PR implements link preview with images. The images are persisted like other assets in the assets database. Tests are added for the green path and to make sure that false image data does not cause upload/persistence. Additionally a test makes sure that if the image upload fails the link preview is still attached and the text message is still send because the text of the message is more important than a little picture of a link preview.

The PR also implements the endpoint for the testservice to have link preview support. The sending of the link preview was tested through the testservice. The receiving of a link preview was not tested and should be part of the implementation in the client.

### CODE REVIEW

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq [hunk 1]
```diff
@@ -42,6 +42,20 @@ CREATE INDEX message_status_index ON Message(status);
 CREATE INDEX message_expire_time_index ON Message(expire_after_millis);
 CREATE INDEX message_self_deletion_end_date ON Message(self_deletion_end_date);
 
+CREATE TABLE MessageLinkPreview (
```
Comment (yamilmedina) [inline persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:45]: I think, we need to include this table in the migration of messages (backup).

`DumpContent.sq`
`ImportContent.sq`

Comment (yamilmedina) [reply persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:45]: Also, this new table needs to have a Migration `sqm` file :D

File: testservice/src/main/kotlin/com/wire/kalium/testservice/models/LinkPreview.kt [hunk 2]
```diff
@@ -18,26 +18,12 @@
...
-    "url": "",
-    "urlOffset": 0
-     */
-}
+data class LinkPreview(
+    val image: LinkPreviewImage? = null,
+    val permanentUrl: String = "",
+    val summary: String = "",
+    val title: String = "",
+    // val tweet = { author: String, username: String }
```
Comment (yamilmedina) [inline testservice/src/main/kotlin/com/wire/kalium/testservice/models/LinkPreview.kt:26]:
```suggestion
    // val tweet = { author: String, username: String }
```

**Review by yamilmedina — CHANGES_REQUESTED** [review]
Thanks very much, finally you hustle to get it done 👏 
I added some comments about missing pieces for migration and backup, that we need when introducing changes in the DB.

As well, since this is targeting 4.8 (release focused on stability) I don't know if it's a good idea to merge it, I'd say it's fine, if we have those migrations at least, and we have enough time to test it 🙌 

Anyway, I discussed with the PM, and we can link it to this ticket, as part of the epic of link previews in Android 
https://wearezeta.atlassian.net/browse/WPB-9877

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq (hunk 1 above)
Comment (mythsunwind) [reply persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:45]: With your help and the help of https://cashapp.github.io/sqldelight/2.0.0-alpha05/multiplatform_sqlite/migrations/ i was able to add the parts for migration and backup. The `verifySqlDelightMigration` task is now successful.

**Review by yamilmedina — APPROVED** [review]

**Comment by yamilmedina** [comment]
Thanks ! for the changes, looks good to me 🙌

**Pull request merged**